pkg/service/controller/servingcert/controller.go

```diff
@@ -0,0 +1,239 @@
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/openshift/origin/pkg/cmd/server/crypto"
+)
+
+const (
+	// ServingCertSecretAnnotation stores the name of the secret to generate into.
+	ServingCertSecretAnnotation = "service.openshift.io/serving-cert-secret-name"
+	// ServingCertCreatedByAnnotation stores the of the signer common name.  This could be used later to see if the
+	// services need to have the the serving certs regenerated.  The presence and matching of this annotation prevents
+	// regeneration
+	ServingCertCreatedByAnnotation = "service.openshift.io/serving-cert-signed-by"
+	// ServingCertErrorAnnotation stores the error that caused cert generation failures.  The presence of this error
+	// prevents retries.
+	ServingCertErrorAnnotation = "service.openshift.io/serving-cert-generation-error"
+	// ServiceUIDAnnotation is an annotation on a secret that indicates which service created it, by UID
+	ServiceUIDAnnotation = "service.openshift.io/originating-service-uid"
+	// ServiceNameAnnotation is an annotation on a secret that indicates which service created it, by Name to allow reverse lookups on services
+	// for comparison against UIDs
+	ServiceNameAnnotation = "service.openshift.io/originating-service-name"
+)
+
+// ServiceServingCertController is responsible for synchronizing Service objects stored
+// in the system with actual running replica sets and pods.
+type ServiceServingCertController struct {
+	serviceClient kclient.ServicesNamespacer
+	secretClient  kclient.SecretsNamespacer
+
+	// Services that need to be checked
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	serviceCache      cache.Store
+	serviceController *framework.Controller
+
+	ca         *crypto.CA
+	publicCert string
+	dnsSuffix  string
+
+	// syncHandler does the work. It's factored out for unit testing
+	syncHandler func(serviceKey string) error
+}
+
+// NewServiceServingCertController creates a new ServiceServingCertController.
+func NewServiceServingCertController(serviceClient kclient.ServicesNamespacer, secretClient kclient.SecretsNamespacer, ca *crypto.CA, dnsSuffix string, resyncInterval time.Duration) *ServiceServingCertController {
+	sc := &ServiceServingCertController{
+		serviceClient: serviceClient,
+		secretClient:  secretClient,
+
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+
+		ca:        ca,
+		dnsSuffix: dnsSuffix,
+	}
+
+	sc.serviceCache, sc.serviceController = framework.NewInformer(
```

Comment:
We may want to share a service informer - we'll have a few in the master.  No change here, but a note.